src/mjolnir/valhalla_openlr_to_edges.cc

```diff
@@ -0,0 +1,368 @@
+  LOG_INFO("Found!");
+  std::ostringstream ss;
+  ss << "start rank: " << start_rank << " end rank: " << end_rank;
+  ss << " lr.getLength: " << lr.getLength();
+  ss << " length: " << length;
+  ss << " edge_ids are: " << boost::algorithm::join(edge_ids, ", ") << std::endl;
+  LOG_INFO(ss.str());
+}
+
+std::vector<std::vector<thor::PathInfo>> compute_path(const baldr::PathLocation& start,
+                                                      const baldr::PathLocation& end,
+                                                      sif::mode_costing_t mode_costing,
+                                                      sif::TravelMode mode,
+                                                      baldr::GraphReader& reader) {
+
+  valhalla::Location origin;
+  valhalla::Location dest;
+
+  baldr::PathLocation::toPBF(start, &origin, reader);
+  baldr::PathLocation::toPBF(end, &dest, reader);
+
+  auto algo = get_path_algorithm(reader, origin, dest);
+  return algo->GetBestPath(origin, dest, reader, mode_costing, mode);
+}
+
+} // namespace
+
+int main(int argc, char** argv) {
+
+  if (argc != 3) {
+    LOG_ERROR("Usage: " + std::string(argv[0]) + " config/file.json openlr_binary");
+    return 1;
+  }
+
+  // config file
+  std::string config_file(argv[1]);
+  boost::property_tree::ptree config;
+  rapidjson::read_json(config_file, config);
+
+  auto reader = std::make_shared<baldr::GraphReader>(config.get_child("mjolnir"));
+
+  valhalla::Options default_options;
+  default_options.set_costing_type(valhalla::Costing::auto_);
+  sif::ParseCosting({}, "", default_options);
+
+  // Construct costing
+  sif::TravelMode mode;
+  auto mode_costing = sif::CostFactory().CreateModeCosting(default_options, mode);
+
+  auto lr = baldr::OpenLR::OpenLr(argv[2], true);
+
+  std::vector<baldr::Location> locations =
+      lr.lrps | ranges::view::transform(lrp_to_location) | ranges::to<std::vector>();
+
+  {
+    // flip the bearing of the last lrp by 180
+    auto& last_location = locations.back();
+    last_location.heading_ = *last_location.heading_ + 180.f;
+    last_location.heading_ = int(*last_location.heading_) % 360;
+  }
+
+  auto projections = loki::Search(locations, *reader, mode_costing[static_cast<size_t>(mode)]);
+
+  assert(lr.lrps.size() == projections.size());
+
+  LOG_INFO("projections size: " + std::to_string(projections.size()));
+
+  auto lrp_ranked_path_edges =
+      std::vector<std::unordered_map<size_t, std::vector<baldr::PathLocation::PathEdge>>>{};
+
+  for (auto const& [idx, lrp] : lr.lrps | ranges::views::enumerate) {
+
+    LOG_INFO("lrp longitude latitude: " + std::to_string(lrp.longitude) + "," +
+             std::to_string(lrp.latitude));
+
+    // filter out the bad projection that are not in accordance with LRP
+    auto rank = [&](const baldr::PathLocation::PathEdge& edge) -> int {
+      // frc mismatch + (motorway mismatch || oneway mismatch || slip road mismatch)
```

Comment:
https://github.com/valhalla/valhalla/issues/4896#issuecomment-2364308539

how am I supposed to test slip road case???

Reply:
slip road is like a ramp or turn channel we marked edges with this property at `directededge::link`